### ISSUE TITLE
pyuwsgi environ handling under macOS

### ISSUE THREAD

**Issue by natefoo** [unbit/uwsgi#1762]
([moved from the mailing list](http://lists.unbit.it/pipermail/uwsgi/2018-March/008870.html))

Because of the... unique... way that macOS handles the standard `environ` [when running as a shared library](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man7/environ.7.html), uWSGI segfaults under certain conditions on macOS when built as a CPython extension (pyuwsgi).

What many projects faced with this issue do is something like:

```c
#if defined(__APPLE__) && defined(UWSGI_AS_SHARED_LIBRARY)
#include <crt_externs.h>
#define environ (*_NSGetEnviron())
#else
extern char **environ;
#endif
```

uWSGI [does something similar](https://github.com/unbit/uwsgi/blob/50ffc6b28a7a84e273fb2b79c8d657b45887fe87/core/uwsgi.c#L29) but instead of the preprocessor define, assigns `environ` to the return of `_NSGetEnviron()` [at runtime](https://github.com/unbit/uwsgi/blob/50ffc6b28a7a84e273fb2b79c8d657b45887fe87/core/uwsgi.c#L2058). However, this approach doesn't work because after a `setenv()`, the address in `environ` is no longer valid, and `_NSGetEnviron()` needs to be called again. That's why the define works, since under that method, `environ` always points at the address returned by `_NSGetEnviron()`.

I made a naive attempt at addressing the problem in #1680 by reinitializing `environ` after the environment is manipulated, but it's not sufficient. Rather than hunting through the code and reinitializing `environ` everywhere, which is sure to be error prone and likely to introduce more bugs in the future, I tried to just switch to using the define method.

Unfortunately, this doesn't work as the `uwsgi_server` and `uwsgi_app` structs have an `environ` member, so there's a name conflict with the defined `environ`. I have two possible solutions but I am not sure how they'd be received by the uWSGI developers, so I wanted to get some guidance before committing any effort:

1. Rename the struct member.
2. Replace access to `environ` throughout the code with a function call returning either `environ` or `_NSGetEnviron()` as appropriate.

The first option seems the most future-proof as it avoids any cases in the future where developers forget to use the function rather than `environ` directly, but it might also be frustrating to have such a core name changed. There are probably simpler solutions, however, as I am by no means a C expert with a lot of tricks up my sleeve.

@unbit's reply on the mailing list:

> By the way, neither of the two solutions will be "zero-cost" (we need to retain ABI backward compatibility) so i think we need to find another way

**Comment by natefoo** [comment]
@unbit I understand why the first idea breaks ABI backward compatibility but I'm not sure why the second does, can you expand on that? I am not fond of my second idea though.

**Comment by pkrog** [comment]
Hi, is someone currently working on this issue?
You hit this issue here at [PhenoMeNal](https://github.com/phnmnl) on our mac computers, and cannot run the latest versions of [Galaxy](https://github.com/galaxyproject).

**Comment by unbit** [comment]
Currently as a super hacky/ugly solution, i suggest to extend the structures with an environ2 (or something similar) field in which (on mac) the environ field will be set (for the other platforms will be a simple copy of the original one)

**Comment by pkrog** [comment]
Thanks @natefoo, I'll try that.

**Comment by natefoo** [comment]
@unbit it looks like `environ` is accessed directly much less than I thought, so fixing it in the few other places where it is, such as in the python plugin loader for `--reload-os-env`, doesn't seem like that bad of a solution after all?

I mention this because even if `environ2` is added, if the `environ` struct member remains, `environ` can't be `#define`d for these other cases globally (such as in `uwsgi.h`).

FWIW, the following patch fixes it for most cases (e.g. if you're not using `--reload-os-env`, I assume):

```patch
diff --git a/core/uwsgi.c b/core/uwsgi.c
index 09da567e..9dd28cff 100644
--- a/core/uwsgi.c
+++ b/core/uwsgi.c
@@ -28,9 +28,10 @@ pid_t masterpid;
 
 #if defined(__APPLE__) && defined(UWSGI_AS_SHARED_LIBRARY)
 #include <crt_externs.h>
-char **environ;
+#define ENVIRON (*_NSGetEnviron())
 #else
 extern char **environ;
+#define ENVIRON environ
 #endif
 
 UWSGI_DECLARE_EMBEDDED_PLUGINS;
@@ -1738,7 +1739,7 @@ static void fixup_argv_and_environ(int argc, char **argv, char **environ, char *
        uwsgi.orig_argv = argv;
        uwsgi.argv = argv;
        uwsgi.argc = argc;
-       uwsgi.environ = environ;
+       uwsgi.environ = ENVIRON;
 
        // avoid messing with fake environ
        if (envp && *environ != *envp) return;
@@ -2055,13 +2056,6 @@ static char *uwsgi_at_file_read(char *filename) {
 }
 
 void uwsgi_setup(int argc, char *argv[], char *envp[]) {
-#ifdef UWSGI_AS_SHARED_LIBRARY
-#ifdef __APPLE__
-       char ***envPtr = _NSGetEnviron();
-       environ = *envPtr;
-#endif
-#endif
-
        int i;
 
        struct utsname uuts;
@@ -2191,7 +2185,7 @@ void uwsgi_setup(int argc, char *argv[], char *envp[]) {
                uwsgi.response_header_limit = UMAX16;
 
        // ok we can now safely play with argv and environ
-       fixup_argv_and_environ(argc, argv, environ, envp);
+       fixup_argv_and_environ(argc, argv, ENVIRON, envp);
 
        if (gethostname(uwsgi.hostname, 255)) {
                uwsgi_error("gethostname()");
@@ -2245,7 +2239,7 @@ void uwsgi_setup(int argc, char *argv[], char *envp[]) {
        build_options();
 #endif
        //parse environ
-       parse_sys_envs(environ);
+       parse_sys_envs(ENVIRON);
 
        // parse commandline options
        uwsgi_commandline_config();
```

**Comment by ipmb** [comment]
I applied the changes from your diff to 2.0.17.1 and it resolved the issue for me.

**Comment by natefoo** [comment]
It looks like @unbit committed this fix in 8022c4e3f60dd8244bf12a723b20d07b82ea0f8d, which is included in 2.0.18. Thanks!
